### ISSUE TITLE
Proxying to an internal resource using `internal` and `proxy` does not serve the correct MIME type

### ISSUE THREAD

**Issue by adamwathan** [mholt/caddy#1086]
### 1. What version of Caddy are you running (`caddy -version`)?

0.9.1

### 2. What are you trying to do?

Proxy access to a static file using the `internal` and `proxy` directives.

### 3. What is your entire Caddyfile?

```text
localhost:2020
internal /css
proxy /redirect http://localhost:8000

```

### 4. How did you run Caddy (give the full command and describe the execution environment)?

`caddy`, on OSX.

### 5. What did you expect to see?

Expected `/redirect` to serve `/css/style.css` with a mime type of `text/css`.

### 6. What did you see instead (give full error messages and/or log)?

`/css/style.css` served with a mime type of `text/html`, causing CSS to not be parsed.

### 7. How can someone who is starting from scratch reproduce this behavior as minimally as possible?

Clone this repository and follow the instructions in the readme:

https://github.com/adamwathan/caddy-mime-bug

**Comment by adamwathan** [comment]
Doing this exact same thing with 0.8.3 does work correctly, for what it's worth.

**Comment by mholt** [comment]
What happens if you add `mime .css text/css` to your Caddyfile? (Sorry I know it's dumb.)

**Comment by adamwathan** [comment]
Still comes back as `text/html`:

![image](https://cloud.githubusercontent.com/assets/4323180/18165929/2e8760e4-7016-11e6-8adb-da20c6aec8c7.png)

**Comment by mholt** [comment]
Oh... that's probably because the URL doesn't have .css in it. Hmmm...

**Comment by adamwathan** [comment]
Yeah I tried to look through the source to see if I could figure out what would need to change to get the old 0.8.3 behavior back but not good enough at Go :)

But to me it sounds like the important thing is to treat the value of the `X-Accel-Redirect` header as the file name, not the URL that then proxied to that file, does that sound right to you?

**Comment by adamwathan** [comment]
If I change the file I'm requesting to be `/redirect/style.css`, then it does load semi-properly but only if I have manually added the mime directive.

It's still giving an error though, because it's being server with _two_ `Content-Type` headers:

![image](https://cloud.githubusercontent.com/assets/4323180/18177829/d75a16fc-7049-11e6-9015-78316ea21332.png)

![image](https://cloud.githubusercontent.com/assets/4323180/18177845/ded8b74e-7049-11e6-9a13-1819883a4063.png)

**Comment by adamwathan** [comment]
I've updated the repo that shows how to reproduce so that you'll see the double header bug when running it:

https://github.com/adamwathan/caddy-mime-bug

**Comment by mholt** [comment]
@adamwathan I finally had a chance to check this out, and I pulled your repro repository here and confirmed that it works as expected. (Maybe you didn't build this branch?)

<img width="938" alt="screen shot 2016-09-24 at 12 14 03 pm" src="https://cloud.githubusercontent.com/assets/1128849/18810320/75c3ebd4-8250-11e6-8983-eca9c50b8e99.png">

Thanks for making that repo, by the way! Made it easy to confirm.

@tw4452852 Thank you for this! I will probably merge soon.

**Comment by adamwathan** [comment]
Ah you're right, I had cloned it down and couldn't get the dependencies installed to build, then pulled it down instead using `go get` and forgot to pull in the branch.

I am still seeing an issue though, it *only* works if you have the `mime` directive specified in the Caddyfile like you mentioned here: 

https://github.com/mholt/caddy/issues/1086#issuecomment-243926768

If you pull down the latest version of that sample repo, you'll see the CSS is coming back as text/html again:

https://github.com/adamwathan/caddy-mime-bug/commit/48333943982fe8cb6e5d447bd7f527b05b30a809

If I pass an empty Content-Type header through, Caddy also preserves that instead of determining the content type itself (not sure what *should* happen here, I just know it happened to work in getting nginx to re-detect the content type on it's own).

Can you think of a way to get Caddy to detect the content type on it's own without having to hardcode the `mime` directive in the Caddyfile? Wouldn't want to do that for this whole list 😳

https://github.com/laravel/valet/blob/master/cli/mimes.php

**Comment by mholt** [comment]
Hmm, well I can tell you that Go won't detect the mime type for ALL of those on its own... but we can probably have it look for the Content-Type as a special case. @tw4452852 what do you think about that? https://golang.org/pkg/net/http/#DetectContentType maybe?

**Comment by adamwathan** [comment]
@mholt Honestly not too worried about catching *all* of them, haha... Whatever I had working in 0.8.3 was fine 😊  

I'm sure it would work well enough if it was possible to tell PHP FPM to send no content type header at all, but unfortunately the best I've been able to do is send an empty Content-Type header, which does trigger FPM not to send the default `text/html` value, but I can't figure out how to suppress it entirely.

**Comment by mholt** [comment]
Man, a lot changed between 0.8.3 and 0.9. I'm not sure which change caused this to happen... 

Wait, why does php send the wrong content-type header?

**Comment by adamwathan** [comment]
It's not that it sends the wrong content type header really, it's that it is trying to say "hey Caddy can you just deal with this?" by using the `X-Accel-Redirect` feature.

I could send the correct mime type by detecting it in PHP but it seems like overkill when everything works fine as long as I send *no* content-type header so Caddy knows to deal with it on it's own.

I did figure out that I can disable the default_mimetype setting in PHP, just not sure what the consequences of that could be. It does make everything work though!

The nice thing about nginx was that I could send a content-type header of `""` (empty string) and nginx would interpret it the same as no content-type header being present at all.

Does that make sense? Haha, sorry for all this man 😬

**Comment by adamwathan** [comment]
Ok so I think I feel confident that in Valet anyways, we can disable that `default_mimetype` setting and everything will be fine using the changes from this PR 👍🏻

The goal of what we were trying to do is send Caddy an HTTP response from the PHP proxy that contained *only* the X-Accel-Redirect header and nothing more, so Caddy would notice and say "Oh PHP wants *me* to take care of this file, no problem."

The problem was that if we didn't send a Content-Type header, PHP FPM was adding a default Content-Type header, and then Caddy was saying "Oh I see you already have a content-type header, I will keep that."

Nginx would let me explicitly send an *empty* Content-Type header, and treat it the same as the header not even being sent (who knows if this is even intentional, I just noticed it worked).

Caddy treats an empty Content-Type header as a header, and preserves the value.

I figured out how to not send the Content-Type header at all, which then triggers Caddy to figure it out, and everything works great.

I *wish* there was a way for me to choose to suppress the default header programmatically, but it seems like I have no choice but to do it PHP-wide at the config level. That sucks, but I *think* it'll still be fine, because Caddy seems to detect `text/html` fine for rendered PHP pages even if PHP doesn't send a content-type along anyways.

**TLDR** with some things tweaked on our side, I think this PR solves our problem. Thanks so much to you guys for helping us on this one!

**Comment by mholt** [comment]
Phew, okay, we'll start by merging that PR and then see about improving it from there. Let me know if it's still not quite right... thanks!

**Comment by mholt** [comment]
We've merged the fix but since it's not *100%* perfect yet I will mark this as deferred so we can remember to pick this up later. Thanks.

**Comment by zhangyoufu** [comment]
Is there any way to keep upstream `Server` header without appending `Server: Caddy`?

**Comment by deloz** [comment]
> Is there any way to keep upstream `Server` header without appending `Server: Caddy`?

You may modify the source.

https://github.com/mholt/caddy/blob/master/caddyhttp/httpserver/server.go#L389
